File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/types/UaBrowseAdditionalInfo.java
package org.opcfoundation.webserver.types;

import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.types.enumerated.BrowseDirection;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.eclipse.milo.opcua.stack.core.types.structured.BrowseDescription;

public class UaBrowseAdditionalInfo {
    private final int maxReferencesPerNode;
    private final int referenceOffset;
    private final int taskCheckListMasks;

    public static final int GET_PARENT_TASK = 1;
    public static final int GET_DEFINITION_TASK = 2;
    public static final int GET_LINK_TASK = 4;

    public static final int GET_CHILD_OBJECT_TASK = 8;
    public static final int GET_CHILD_VARIABLE_TASK = 16;
    public static final int GET_CHILD_METHOD_TASK = 32;
    public static final int GET_CHILD_TASK = GET_CHILD_OBJECT_TASK + GET_CHILD_VARIABLE_TASK + GET_CHILD_METHOD_TASK;

    public static final int ALL_TASK = GET_PARENT_TASK + GET_CHILD_TASK + GET_DEFINITION_TASK + GET_LINK_TASK;

    public UaBrowseAdditionalInfo(
            int maxReferencesPerNode,
            int referenceOffset,
            int taskCheckListMasks)
    {
        this.maxReferencesPerNode = Math.max(maxReferencesPerNode, 0);
        this.referenceOffset = Math.max(referenceOffset, 0);
        this.taskCheckListMasks = taskCheckListMasks;
    }

    public int getMaxReferencesPerNode() {
        return maxReferencesPerNode;
    }

    public int getReferenceOffset() {
        return referenceOffset;
    }

    public int getTaskCheckListMasks() {
        return taskCheckListMasks;
    }

    public UaBrowseAdditionalInfo updateOffset(int offsetToAdd)
    {
        return new UaBrowseAdditionalInfo(maxReferencesPerNode, referenceOffset + offsetToAdd, taskCheckListMasks);
    }

    public UaBrowseAdditionalInfo updateTasks(BrowseDescription description)
    {
        if (taskCheckListMasks != 0) return this;

        int browseInfoMask = 0;

        if (description.getIncludeSubtypes())
        {
            if ((description.getReferenceTypeId().equals(NodeIds.References) ||
                    description.getReferenceTypeId().equals(NodeIds.HierarchicalReferences) ||
                    description.getReferenceTypeId().equals(NodeIds.Aggregates)))
            {
                if (description.getBrowseDirection() == BrowseDirection.Inverse ||
                        description.getBrowseDirection() == BrowseDirection.Both)
                {
                    if ((description.getNodeClassMask().longValue() & NodeClass.Object.getValue()) != 0) {
                        browseInfoMask = browseInfoMask | GET_PARENT_TASK; // Get parent
                    }
                }

                if (description.getBrowseDirection() == BrowseDirection.Forward ||
                        description.getBrowseDirection() == BrowseDirection.Both) {
                    if ((description.getNodeClassMask().longValue() & NodeClass.Object.getValue()) != 0) {
                        browseInfoMask = browseInfoMask | GET_CHILD_OBJECT_TASK;  // Get object child
                    }

                    if ((description.getNodeClassMask().longValue() & NodeClass.Variable.getValue()) != 0) {
                        browseInfoMask = browseInfoMask | GET_CHILD_VARIABLE_TASK;  // Get variable child
                    }

                    if ((description.getNodeClassMask().longValue() & NodeClass.Method.getValue()) != 0) {
                        browseInfoMask = browseInfoMask | GET_CHILD_METHOD_TASK;  // Get method child
                    }
                }
            }

            if (description.getReferenceTypeId().equals(NodeIds.References) ||
                    description.getReferenceTypeId().equals(NodeIds.NonHierarchicalReferences))
            {
                if ((description.getNodeClassMask().longValue() & NodeClass.ObjectType.getValue()) != 0)
                {
                    browseInfoMask = browseInfoMask | GET_DEFINITION_TASK; // Get type definition
                }

                if ((description.getNodeClassMask().longValue() & NodeClass.Object.getValue()) != 0)
                {
                    browseInfoMask = browseInfoMask | GET_LINK_TASK; // Get links
                }
            }
        } else {
            if (description.getBrowseDirection() == BrowseDirection.Forward ||
                    description.getBrowseDirection() == BrowseDirection.Both)
            {
                if ((description.getNodeClassMask().longValue() & NodeClass.Object.getValue()) != 0)
                {
                    if (description.getReferenceTypeId().equals(NodeIds.HasComponent))
                    {
                        browseInfoMask = browseInfoMask | GET_CHILD_OBJECT_TASK;  // Get object child
                    }
                }

                if ((description.getNodeClassMask().longValue() & NodeClass.Variable.getValue()) != 0)
                {
                    if (description.getReferenceTypeId().equals(NodeIds.HasComponent) ||
                            description.getReferenceTypeId().equals(NodeIds.HasProperty))
                    {
                        browseInfoMask = browseInfoMask | GET_CHILD_VARIABLE_TASK;  // Get variable child
                    }
                }

                if ((description.getNodeClassMask().longValue() & NodeClass.Method.getValue()) != 0)
                {
                    if (description.getReferenceTypeId().equals(NodeIds.HasComponent))
                    {
                        browseInfoMask = browseInfoMask | GET_CHILD_METHOD_TASK;  // Get method child
                    }
                }

                if ((description.getNodeClassMask().longValue() & NodeClass.ObjectType.getValue()) != 0 &&
                        description.getReferenceTypeId().equals(NodeIds.HasTypeDefinition))
                {
                    browseInfoMask = browseInfoMask | GET_DEFINITION_TASK; // Get type definition
                }
            }
        }

        return new UaBrowseAdditionalInfo(maxReferencesPerNode, referenceOffset, browseInfoMask);
    }

    public boolean isTaskRequired(int taskMask)
    {
        return (taskCheckListMasks & taskMask) != 0;
    }

    public UaBrowseAdditionalInfo taskComplete(int taskMask)
    {
        if (!isTaskRequired(taskMask)) return this;

        return new UaBrowseAdditionalInfo(
                maxReferencesPerNode,
                0,
                taskCheckListMasks&(~taskMask));
    }

    public boolean isAllTaskComplete()
    {
        return (0 == taskCheckListMasks);
    }

    @Override
    public String toString()
    {
        String ret = "maxReferencesPerNode: " + maxReferencesPerNode;
        ret += " referenceOffset: " + referenceOffset;
        ret += " taskCheckListMasks: " + taskCheckListMasks;
        return ret;
    }
}

File: opcua-webservice-node/src/server/types/common/UaBrowseAdditionalInfo.ts
import { BrowseDirection, NodeClass } from "opcua-webapi";
import { ReferenceTypeIds, UaBrowseDescription, UaNodeId } from "opcua-webapi-ts";

export class UaBrowseAdditionalInfo {
    static readonly GET_PARENT_TASK = 1;
    static readonly GET_DEFINITION_TASK = 2;
    static readonly GET_LINK_TASK = 4;

    static readonly GET_CHILD_OBJECT_TASK = 8;
    static readonly GET_CHILD_VARIABLE_TASK = 16;
    static readonly GET_CHILD_METHOD_TASK = 32;
    static readonly GET_CHILD_TASK =
        UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK +
        UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK +
        UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK;

    static readonly ALL_TASK =
        UaBrowseAdditionalInfo.GET_PARENT_TASK +
        UaBrowseAdditionalInfo.GET_CHILD_TASK +
        UaBrowseAdditionalInfo.GET_DEFINITION_TASK +
        UaBrowseAdditionalInfo.GET_LINK_TASK;

    private readonly _maxReferencesPerNode: number;
    private readonly _referenceOffset: number;
    private readonly _taskCheckListMasks: number;

    constructor(
        maxReferencesPerNode: number,
        referenceOffset: number,
        taskCheckListMasks: number,
    ) {
        this._maxReferencesPerNode = Math.max(maxReferencesPerNode, 0);
        this._referenceOffset = Math.max(referenceOffset, 0);
        this._taskCheckListMasks = taskCheckListMasks;
    }

    get maxReferencesPerNode(): number {
        return this._maxReferencesPerNode;
    }

    get referenceOffset(): number {
        return this._referenceOffset;
    }

    get taskCheckListMasks(): number {
        return this._taskCheckListMasks;
    }

    updateOffset(offsetToAdd: number): UaBrowseAdditionalInfo {
        return new UaBrowseAdditionalInfo(
            this._maxReferencesPerNode,
            this._referenceOffset + offsetToAdd,
            this._taskCheckListMasks,
        );
    }

    updateTasks(description: UaBrowseDescription): UaBrowseAdditionalInfo {
        if (this._taskCheckListMasks !== 0) {
            return this;
        }

        let browseInfoMask = 0;
        const referenceTypeId = description.referenceTypeId;
        const browseDirection = description.browseDirection;
        const includeSubtypes = description.includeSubtypes;
        const nodeClassMask = description.nodeClassMask ?? 0;

        if (includeSubtypes) {
            if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.References)) ||
                referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HierarchicalReferences)) ||
                referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.Aggregates))) {
                if ((browseDirection === BrowseDirection.Inverse || browseDirection === BrowseDirection.Both) &&
                    this.hasNodeClass(nodeClassMask, NodeClass.Object)
                ) {
                    browseInfoMask |= UaBrowseAdditionalInfo.GET_PARENT_TASK;
                }

                if (browseDirection === BrowseDirection.Forward || browseDirection === BrowseDirection.Both) {
                    if (this.hasNodeClass(nodeClassMask, NodeClass.Object)) {
                        browseInfoMask |= UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK;
                    }

                    if (this.hasNodeClass(nodeClassMask, NodeClass.Variable)) {
                        browseInfoMask |= UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK;
                    }

                    if (this.hasNodeClass(nodeClassMask, NodeClass.Method)) {
                        browseInfoMask |= UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK;
                    }
                }
            }

            if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.References)) ||
                referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.NonHierarchicalReferences))) {
                if (browseDirection === BrowseDirection.Forward || browseDirection === BrowseDirection.Both) {
                    if (this.hasNodeClass(nodeClassMask, NodeClass.ObjectType)) {
                        browseInfoMask |= UaBrowseAdditionalInfo.GET_DEFINITION_TASK;
                    }

                    if (this.hasNodeClass(nodeClassMask, NodeClass.Object)) {
                        browseInfoMask |= UaBrowseAdditionalInfo.GET_LINK_TASK;
                    }
                }
            }
        } else if (browseDirection === BrowseDirection.Forward || browseDirection === BrowseDirection.Both) {
            if (this.hasNodeClass(nodeClassMask, NodeClass.Object) &&
                referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasComponent))) {
                browseInfoMask |= UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK;
            }

            if (this.hasNodeClass(nodeClassMask, NodeClass.Variable) &&
                (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasComponent)) ||
                    referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasProperty)))) {
                browseInfoMask |= UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK;
            }

            if (this.hasNodeClass(nodeClassMask, NodeClass.Method) &&
                referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasComponent))) {
                browseInfoMask |= UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK;
            }

            if (this.hasNodeClass(nodeClassMask, NodeClass.ObjectType)) {
                if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasTypeDefinition))) {
                    browseInfoMask |= UaBrowseAdditionalInfo.GET_DEFINITION_TASK;
                }

                if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.GeneratesEvent))) {
                    browseInfoMask |= UaBrowseAdditionalInfo.GET_LINK_TASK;
                }
            }
        }

        return new UaBrowseAdditionalInfo(
            this._maxReferencesPerNode,
            this._referenceOffset,
            browseInfoMask,
        );
    }

    isTaskRequired(taskMask: number): boolean {
        return (this._taskCheckListMasks & taskMask) !== 0;
    }

    taskComplete(taskMask: number): UaBrowseAdditionalInfo {
        if (!this.isTaskRequired(taskMask)) {
            return this;
        }

        return new UaBrowseAdditionalInfo(
            this._maxReferencesPerNode,
            0,
            this._taskCheckListMasks & ~taskMask
        );
    }

    isAllTaskComplete(): boolean {
        return this._taskCheckListMasks === 0;
    }

    private hasNodeClass(nodeClassMask: number, nodeClass: NodeClass): boolean {
        return (nodeClassMask & nodeClass) !== 0;
    }
}